designs/2022-languages/README.md

```diff
@@ -0,0 +1,723 @@
+- Repo: eslint/eslint
+- Start Date: 2022-11-25
+- RFC PR: https://github.com/eslint/rfcs/pull/99
+- Authors: Nicholas C. Zakas
+
+# ESLint Language Plugins
+
+## Summary
+
+This proposal provides a way for plugin authors to define parsers, rules, etc., for any arbitrary programming languages. This would allow ESLint to work on any type of programming language for which there is a plugin defined, not just for JavaScript and its variants.
+
+## Motivation
+
+Over the nine years ESLint has been in existence, it has grown into being not just the most popular JavaScript linter, but arguably the most popular linter for any programming language. Twitter is filled with people lamenting that they don't have an "ESLint for X", where X is some other file type they're working with. Numerous projects have popped up that mimic a lot of what ESLint does at it's core:
+
+* [Stylelint](https://stylelint.io/);
+* [Markdownlint](https://github.com/DavidAnson/markdownlint)
+
+In addition, there are numerous projects that hack around other languages to make them work in ESLint, such as:
+
+* [GraphQL-ESLint](https://github.com/B2o5T/graphql-eslint)
+* [eslint-plugin-json](https://www.npmjs.com/package/eslint-plugin-json)
+* [typescript-eslint](https://typescript-eslint.io)
+
+Instead of forcing others to rewrite what is core linting functionality, or asking people to produce an ESTree format AST for some language other than JavaScript, we can make everyone's lives easier by providing formal language support such that languages can be plugged in to ESLint easily.
+
+We did also previously accept [#56](https://github.com/eslint/rfcs/pull/56) as an extension to `parseForESLint()` to support more languages and [eslint/eslint#14745](https://github.com/eslint/eslint/issues/14745) has been open for a couple of years to figure out a way to lint both a file and its virtual parts.
```

Comment:
> > How would user configure ESLint so JS file is parsed correctly and run rules both for JS and for CSS-in-JS on a single file?
> 
> It sounds like you are saying that you parse the same file into two different ASTs, is that correct? If so, it might be better to use a [processor](https://eslint.org/docs/latest/extend/custom-processors) instead of a language. Processors are designed for this use case where you might want to slice up a file into different types of files and have ESLint work on those individually.

@nzakas I think https://github.com/eslint/eslint/issues/14745 is the key to what I asked.

Processors are limited to one per file, if I understand correctly. If we would have a JS file with CSS and GraphQL inside, we would be able to use either processor to get CSS or processor to get GraphQL, but not both. Which is not what user would want. User would want to lint all three languages for a single file: JS, CSS, and GraphQL.

Reply:
This isn't strictly true...there are definitely ways to accomplish what you want, but they are complicated. I'd like to make it simpler but perhaps this RFC isn't the right spot for that. (See my last comments on the RFC as a whole.)